CONTRIBUTING.rst

```diff
@@ -1,107 +1,414 @@
+|                    |             | - No new dependency                                |
+|                    |             | - Backward compatibility                           |
++--------------------+-------------+----------------------------------------------------+
+|                    |             | - Test type is adapted to function behavior        |
+|                    |             | - Tests pass continuous integration                |
+|                    |  Bugfixes   | - Coverage doesn't decrease                        |
+|      `Tests`_      | New features| - Fast, using small mocked data                    |
+|                    |             | - Atomic (one per function) and seeded             |
+|                    |             | - For Bugfixes: non-regression test                |
++--------------------+-------------+----------------------------------------------------+
+|                    |             | - Clearly showcase benefits                        |
+|      Examples      | New features| - Run in a few seconds                             |
+|                    |             | - Use light data (generated or from Nilearn)       |
+|                    |             | - Renders well after build                         |
++--------------------+-------------+----------------------------------------------------+
+|                    |             | - Simple and didactic                              |
+|  `Documentation`_  |    Any      | - Links to relevant examples                       |
+|                    |             | - Renders well after build                         |
+|                    |             | - Doesn't include code                             |
++--------------------+-------------+----------------------------------------------------+
+
+PR Structure
+-------------
+
+A new pull request must have a clear scope, conveyed through its name, a
+reference to the issue it targets (through the exact mention "Closes #XXXX")
+as well as a synthetic summary of its goals and main steps.
+When working on big contributions, we advise contributors to split them in
+several PRs when possible.
+This has the benefit to make code changes clearer, easier to make and to review
+and overall smoothen the whole process.
+No changes unrelated to the PR should be included.
+
+When relevant, PR names should also include tags if they fall in various
+categories.
+If the authors feels he still has work to do before a first round of review, he
+can include the [WIP] tag or conversely, the [MRG] tag to highlight this PR is
+ready for review.
+Other tags can describe the PR content : [FIX] for a bugfix, [DOC] for a
+change in documentation or examples, [ENH] for a new feature and [MAINT] for
+maintenance changes.
+
+Coding Style
+-------------
+
+The main conventions we follow are : line length < 80, spaces around operators,
+variable names, function names are underscore separated (a_nice_function),
```

Comment:
```suggestion
variable names, function names are underscore separated (e.g., ``a_nice_function``) and as short as possible,
public functions exposed in their parent module's init file, private function names preceded with a "_",
```

These are from @NicolasGensollen's https://github.com/nilearn/nilearn/pull/2755/files#r703577151.

@NicolasGensollen I wanted to include "Functions should avoid side effects (or it should be stated clearly in the docstring)" here, but I'm not sure what you mean.

Reply:
@tsalo I was basically saying to favor pure functions as much as possible. 
For example, avoid functions modifying their inputs.

Reply:
Ohhh gotcha. Agreed, that's definitely worth mentioning.